src/signer-pool.ts

```diff
@@ -0,0 +1,50 @@
+import { ethers } from "ethers";
+import { Logger } from "winston";
+import { createLogger } from "./logging";
+
+class SignerPool {
+  signers: ethers.Signer[];
+  pool: number[];
+  logger: Logger;
+  constructor(signers: ethers.Signer[]) {
+    this.signers = signers;
+    this.pool = Array.from(Array(this.signers.length).keys());
+    // this.acquireCallbacks = []; unused
+    this.logger = createLogger({ componentName: "SignerPool" });
+  }
+
+  static async create({ signers }: { signers: ethers.Signer[] }) {
+    return new SignerPool(signers);
+  }
+
+  async acquire(): Promise<[number, ethers.Signer] | undefined> {
+    this.logger.info("awaiting signer");
+    while (!this.pool.length) {
+      await new Promise((resolve, reject) => setTimeout(resolve, 1));
+    }
+    const i = this.pool.pop();
+    if (i === undefined) return undefined;
+    this.logger.info(`acquired signer i=${i}`);
+    return [i, this.signers[i]];
+  }
+
+  release(i: number) {
+    this.logger.info(`released signer i=${i}`);
+    this.pool.push(i);
+  }
+
+  async withSigner(cb: (signer: ethers.Signer) => Promise<void>) {
+    const [i, signer] = (await this.acquire()) || [];
+    if (!signer || i === undefined)
+      return Promise.reject("No signer available");
```

Comment:
Once [this](https://github.com/Synthetixio/futures-keepers/pull/7/commits/91cbfe7199d6edd4f79dac0aa91d84c88ec761bf#r773523069) comment is addressed we can remove this check